WordPressVIPMinimum/Sniffs/Security/ProperEscapingFunctionSniff.php

```diff
@@ -53,8 +54,27 @@ public function process_token( $stackPtr ) {
 
 		$function_name = $this->tokens[ $stackPtr ]['content'];
 
-		$echo_or_string_concat = $this->phpcsFile->findPrevious( Tokens::$emptyTokens, $stackPtr - 1, null, true );
+		$data = [ $function_name ];
 
+		if ( $function_name === 'esc_attr' || $function_name === 'esc_attr__' ) {
+			// Find esc_attr being used between HTML tags.
+			$tokens = array_merge(
+				Tokens::$emptyTokens,
+				[
+					'T_ECHO'     => T_ECHO,
+					'T_OPEN_TAG' => T_OPEN_TAG,
+				]
+			);
+			$html_tag = $this->phpcsFile->findPrevious( $tokens, $stackPtr - 1, null, true );
+			if ( $this->tokens[ $html_tag ]['type'] === 'T_INLINE_HTML' && false !== strpos( $this->tokens[ $html_tag ]['content'], '>' ) ) {
```

Comment:
* Why only look at inline HTML and not at other text strings, like `T_CONSTANT_ENCAPSED_STRING` ?
    Consider the following code:
    ```php
    echo '<h1>' . esc_attr__( $title, 'domain' ) . '</h1>';
    echo '<h1>', esc_attr__( $title, 'domain' ), '</h1>';
    ```
    Note: the second example passes comma separated arguments to `echo`. This is valid PHP (and more performant than concatenation).
* Is this `strpos()` correct ? Is `strpos()` the right function to use ?
    Consider the following code:
    ```php
    <h1><span class="<?php echo esc_attr__( $title, 'domain' ); ?>">Text</span></h1>
    ```

Nitpick: we moved away from Yoda conditions, Please don't re-introduce them.

Reply:
I think `substr()` would be a better one to use over `strpos()` since we only care about the `>` being on the last part of the stack pointer content (if it's a `T_CONSTANT_ENCAPSED_STRING` we might have to just strip out any trailing quotes).

The Yoda conditions are a bad habit, must have sneaked in unwillingly 😄.

TIL about `,` being used as a concatenation operator!

Reply:
@jrfnl Thank you for the useful feedback, I've iterated on it! 🙂

Why only look at inline HTML and not at other text strings, like T_CONSTANT_ENCAPSED_STRING ?

Do you think I need to expand the sniff to account for more text string tokens (besides T_CONSTANT_ENCAPSED_STRING) such as the ones listed in $textStringTokens? https://github.com/squizlabs/PHP_CodeSniffer/blob/8b0137a78cfae9b8fe1c8afdf5396c8a8d6a25af/src/Util/Tokens.php#L537?

Reply:
> TIL about , being used as a concatenation operator!

Sorry for confusing you. The `,` is *not* a concatenation operator. `echo` as a language construct, however, allows for passing multiple "arguments" which will then each be echo-ed out. The net effect on the output is the same, but it can save a lot of memory using comma's instead of concatenation.

As per the [documentation](https://www.php.net/manual/en/function.echo.php):
>  Outputs all parameters. No additional newline is appended.
> 
> `echo` is not actually a function (it is a language construct), so you are not required to use parentheses with it. `echo` (unlike some other language constructs) does not behave like a function, so it cannot always be used in the context of a function. Additionally, if you want to pass more than one parameter to `echo`, the parameters must not be enclosed within parentheses. 

Note: this only works for `echo`, not for `print`.

Reply:
> I think substr() would be a better one to use over strpos() since we only care about the `>` being on the last part of the stack pointer content (if it's a T_CONSTANT_ENCAPSED_STRING we might have to just strip out any trailing quotes).

Sounds like a better approach.

> Do you think I need to expand the sniff to account for more text string tokens (besides T_CONSTANT_ENCAPSED_STRING) such as the ones listed in $textStringTokens? 

Well, heredoc and nowdoc do not allow for function calls within the text block. Even so, that shouldn't be an issue as that just means those tokens would (should) never match. So, yes, using `Tokens::$textStringTokens` will probably be a better approach.